arch/aarch64/fill_window_arm.c

```diff
@@ -16,6 +17,46 @@ extern ZLIB_INTERNAL int read_buf        (z_stream *strm, unsigned char *buf, un
 
 #if __ARM_NEON
 #include <arm_neon.h>
+
+/* SIMD version of hash_chain rebase */
+static inline void slide_hash_chain(Pos *table, int entries, uint16_t window_size)
+{
+    register uint16x8_t v, *p;
+    register size_t n;
+
+    size_t size = entries*sizeof(table[0]);
+    Assert((size % sizeof(int16x8_t) * 8 == 0), "hash table size err");
+
+    Assert(sizeof(Pos) == 2, "Wrong Pos size");
+    v = vdupq_n_u16(window_size);
+
+    p = (uint16x8_t *)table;
+    n = size / (sizeof(uint16x8_t) * 8);
+    do {
+        p[0] = vqsubq_u16(p[0], v);
+        p[1] = vqsubq_u16(p[1], v);
+        p[2] = vqsubq_u16(p[2], v);
+        p[3] = vqsubq_u16(p[3], v);
+        p[4] = vqsubq_u16(p[4], v);
+        p[5] = vqsubq_u16(p[5], v);
+        p[6] = vqsubq_u16(p[6], v);
+        p[7] = vqsubq_u16(p[7], v);
+        p += 8;
+    } while (--n);
+}
+#else
+/* generic version for hash rebase */
+static inline void slide_hash_chain(Pos *table, int entries, uint16_t window_size)
+{
+    int n = entries;
+    Pos *p = &table[n];
+    unsigned m;
+
+    do {
+        m = *--p;
+        *p = (Pos)(m >= window_size ? m-window_size : NIL);
+    } while (--n);
```

Comment:
This code is a lot longer in assembly than the original for-loop. This version was used in original zlib because of gcc bug when compiling for x86_64.